### ISSUE TITLE
Jaccard Index behaviour when num_classes is absent

### ISSUE THREAD

**Issue by drozzy** [PyTorchLightning/metrics#829]
The [jaccard index function](https://github.com/PyTorchLightning/metrics/blob/8fade87062a7b87c1e6429bbe1c4e0112b3713a5/torchmetrics/functional/classification/jaccard.py#L69) determines the number of classes automatically.
This can result in strange results if the user forgets to specify `num_classes`. For example, in IoU determination in semantic segmentation, not all images have all classes. So determining them automatically will almost always lead to wrong number.
Additionally, the [warning](https://github.com/PyTorchLightning/metrics/blob/8fade87062a7b87c1e6429bbe1c4e0112b3713a5/torchmetrics/utilities/data.py#L156) that shows up makes it seem like something is wrong, when the user specifies the number of classes explicitly (but this is not the case at all):

```python
/run/determined/pythonuserbase/lib/python3.7/site-packages/torchmetrics/utilities/prints.py:36: RuntimeWarning: You have set 11 number of classes which is different from predicted (1000) and target (10) number of classes
```

Would it be crazy if `num_classes` requirement is default and the automatic determination of classes is a special option that must be enabled?

**Comment by SkafteNicki** [comment]
Its fine by me to update this to be a required argument.

**Comment by Borda** [comment]
@drozzy mind sending a PR? :rabbit:

**Comment by SkafteNicki** [comment]
Should be fixed in PR #853
